make tests not dependent to an external server

kpouer opened the pull request:
Hi, a few doctests were dependent on a SMPP server which is not accessible (at least publicly). This pull request is to launch a local SMPP server instead to pass the test.
Also, it seems the test is the same in lib.rs and command_codec.rs. Maybe there is a reason to have both, otherwise one of them could be removed.
Regards

Matthieu

kpouer commented:
It seems you changed a little the doctest at the same time I wrote this pull request, I will take a look at it to fix the pull request this afternoon

JadKHaddad commented:
@kpouer We need an [AsyncRead](https://docs.rs/tokio/latest/tokio/io/trait.AsyncRead.html) and an [AsyncWrite](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWrite.html) for the [Framed](https://docs.rs/tokio-util/latest/tokio_util/codec/struct.Framed.html). So I was thinking about using a [DuplexStream](https://docs.rs/tokio/latest/tokio/io/struct.DuplexStream.html) to prevent parallel tests from binding on the same port.

kpouer commented:
Oh you are right about async, I will update it and was not aware of DuplexStream, I will take a look at it too

kpouer commented:
I update the PR, what it what you had in mind ? I also added a few println even if doctests do not show them, maybe Rust will change this behavior. And added the code in samples to launch it and see logs as well

JadKHaddad commented on examples/client_server.rs:
```diff
@@ -0,0 +1,52 @@
+/// Demonstrate a SMPP server and a client sending an EnquireLink
```
Can we use `module-level` documentation here?. AKA `//!`

kpouer replied:
why not, however I am not aware about the difference, what will it do  in that case ?

JadKHaddad commented on examples/client_server.rs:
```diff
@@ -0,0 +1,52 @@
+/// Demonstrate a SMPP server and a client sending an EnquireLink
+/// This example requires `tokio-codec` feature
+
```
And some instructions on how to run the example. Like [submit_sm](https://github.com/JadKHaddad/Rusmpp/blob/main/examples/submit_sm.rs#L4).

JadKHaddad commented on examples/client_server.rs:
For this example to work, we have to add it to `Cargo.toml`, like [example](https://github.com/JadKHaddad/Rusmpp/blob/main/Cargo.toml#L31). Tracing feature is not needed tho.

JadKHaddad commented on examples/client_server.rs:
```diff
@@ -0,0 +1,52 @@
+use tokio::io::DuplexStream;
+use tokio_util::codec::Framed;
+
+async fn launch_server(server_stream: DuplexStream) -> Result<(), Box<dyn std::error::Error>> {
```
`Framed` needs an `AsyncRead + AsyncWrite` and `DuplexStream` is `AsyncRead + AsyncWrite`. But making this function take a concrate type: `DuplexStream`,  might confuse new users. Maybe we can use `Generics`:

```rust
async fn launch_server<S: AsyncRead + AsyncWrite>(server_stream: S) {
}
```

JadKHaddad commented on examples/client_server.rs:
So I came up with this:

```rust
use futures::{SinkExt, StreamExt};
use rusmpp::{
    codec::command_codec::CommandCodec,
    commands::{
        command::Command,
        pdu::Pdu,
        types::{command_id::CommandId, command_status::CommandStatus},
    },
};
use tokio::io::{AsyncRead, AsyncWrite};
use tokio_util::codec::Framed;

async fn run_server<S: AsyncRead + AsyncWrite + Unpin>(
    stream: S,
) -> Result<(), Box<dyn std::error::Error>> {
    let mut framed = Framed::new(stream, CommandCodec {});

    while let Some(Ok(command)) = framed.next().await {
        if let CommandId::EnquireLink = command.command_id() {
            println!("Server: EnquireLink received");

            let response = Command::builder()
                .command_status(CommandStatus::EsmeRok)
                .sequence_number(command.sequence_number)
                .pdu(Pdu::EnquireLinkResp)
                .build();

            framed.send(&response).await?;

            println!("Server: EnquireLink response sent");

            break;
        }
    }

    Ok(())
}

async fn run_client<S: AsyncRead + AsyncWrite + Unpin>(
    stream: S,
) -> Result<(), Box<dyn std::error::Error>> {
    let mut framed = Framed::new(stream, CommandCodec {});

    let enquire_link_command = Command::new(CommandStatus::EsmeRok, 0, Pdu::EnquireLink);

    framed.send(&enquire_link_command).await?;

    println!("Client: EnquireLink sent");

    while let Some(Ok(command)) = framed.next().await {
        if let CommandId::EnquireLinkResp = command.command_id() {
            println!("Client: EnquireLink response received");

            break;
        }
    }

    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let (server_stream, client_stream) = tokio::io::duplex(4096);

    let (server_result, client_result) =
        tokio::join!(run_server(server_stream), run_client(client_stream));

    server_result?;
    client_result?;

    Ok(())
}
```
I think, doing it like this makes it more clear, what do you think?

kpouer replied:
I have some doubts, doing this the server cannot run in a different thread. It works for the test but if you write a real server you might wan't to use a thread to run the smpp server and use the main thread for other purpose ?

JadKHaddad reviewed (commented):
Do not forget to run `cargo fmt` ;)

JadKHaddad commented:
@kpouer `println!` can be removed in the doctests, as they do not provide any benefits. But they can stay in the example :)
